Finalize postMessage support to being created service worker clients

youennf opened the pull request:
#### 8f8c1da762686cec4b5140d95b4212945cdd75bb
<pre>
Finalize postMessage support to being created service worker clients
<a href="https://bugs.webkit.org/show_bug.cgi?id=255203">https://bugs.webkit.org/show_bug.cgi?id=255203</a>
rdar://problem/107802408

Reviewed by Chris Dumez.

We change the way service worker client postMessage is done to fully align with the spec.
Whenever posting a message to a service worker client, we go to network process.
Previously, we were directly going to the client web process.

We are now checking whether the client is a being created client.
A being created client is a client registered in network process when receiving a navigation or main script load.
The client, when fully created, is updating its data by registering itself through IPC.
At this point, the client is no longer a being created client.
A client is a being created client if SWServer:: m_clientPendingMessagesById has an entry with the client ID.

As long as the client is a being created client, we store the message until the client is created in the web process.
When the client is registering itself, we remove the entry in SWServer:: m_clientPendingMessagesById and send all pending messages.

This allows direct support of workers since we reuse the normal post message process.
This also gets us closer to spec, in  particular:
- We ensure message ordering is always preserved.
- We ensure we do not overbuffer messages.

Covered by rebased newly added WPT tests.

* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/controlled-dedicatedworker-postMessage.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/controlled-iframe-postMessage.https-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::m_isNonRenderedPlaceholder):
* Source/WebCore/workers/service/SWClientConnection.cpp:
(WebCore::postMessageToContainer):
(WebCore::SWClientConnection::postMessageToServiceWorkerClient):
* Source/WebCore/workers/service/SWClientConnection.h:
* Source/WebCore/workers/service/ServiceWorkerContainer.cpp:
(WebCore::ServiceWorkerContainer::ServiceWorkerContainer):
(WebCore::ServiceWorkerContainer::startMessages):
(WebCore::ServiceWorkerContainer::postMessage):
(WebCore::ServiceWorkerContainer::addEventListener):
* Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
(WebCore::WorkerSWClientConnection::getServiceWorkerClientPendingMessages): Deleted.
* Source/WebCore/workers/service/WorkerSWClientConnection.h:
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::registerServiceWorkerClient):
(WebCore::SWServer::postMessageToServiceWorkerClient):
(WebCore::SWServer::addServiceWorkerClientPendingMessage): Deleted.
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
(WebKit::WebSWServerConnection::controlClient):
(WebKit::WebSWServerConnection::postMessageToServiceWorkerClient):
(WebKit::WebSWServerConnection::registerServiceWorkerClient):
(WebKit::WebSWServerConnection::registerServiceWorkerClientInternal):
(WebKit::WebSWServerConnection::getServiceWorkerClientPendingMessages): Deleted.
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.messages.in:
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp:
(WebKit::WebSWClientConnection::getServiceWorkerClientPendingMessages): Deleted.
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.h:
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.messages.in:

Canonical link: <a href="https://commits.webkit.org/262818@main">https://commits.webkit.org/262818@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/873893021199811febba831780c80449fb321690

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2691 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2692 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2833 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4099 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3062 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2655 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2835 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2789 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2368 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2716 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/3160 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2407 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3863 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/654 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2392 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/2254 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2774 "Built successfully") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2410 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3621 "Passed tests") | 
| | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2443 "Passed tests") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/2223 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2443 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2395 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/663 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2434 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2587 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/workers/service/server/SWServer.h:
```diff
@@ -213,7 +213,8 @@ class SWServer : public CanMakeWeakPtr<SWServer> {
     WEBCORE_EXPORT static HashSet<SWServer*>& allServers();
 
-    WEBCORE_EXPORT void registerServiceWorkerClient(ClientOrigin&&, ServiceWorkerClientData&&, const std::optional<ServiceWorkerRegistrationIdentifier>&, String&& userAgent);
+    enum class IsBeingCreatedClient { No, Yes };
```
: bool

youennf replied:
ok

cdumez commented on Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
```diff
@@ -386,17 +387,8 @@ void WebSWServerConnection::scheduleUnregisterJobInServer(ServiceWorkerJobIdenti
-
-        server().addServiceWorkerClientPendingMessage(destinationContextIdentifier, { message, sourceServiceWorkerData, sourceOrigin });
+    server().postMessageToServiceWorkerClient(destinationContextIdentifier, message, sourceIdentifier, sourceOrigin, [this] (auto destinationContextIdentifier, auto& message, auto sourceServiceWorkerData, auto& sourceOrigin) {
+        send(Messages::WebSWClientConnection::PostMessageToServiceWorkerClient { destinationContextIdentifier, message, sourceServiceWorkerData, sourceOrigin }, 0);
```
What guarantees that `this` is still alive here?

youennf replied:
We pass a const Function& so that can only be called synchronously.
I probably need to refactor a bit this code anyway to handle the case of postMessaging to a PSON service worker client. This will probably remove this lambda.

youennf committed:
Finalize postMessage support to being created service worker clients
https://bugs.webkit.org/show_bug.cgi?id=255203
rdar://problem/107802408

Reviewed by Chris Dumez.

We change the way service worker client postMessage is done to fully align with the spec.
Whenever posting a message to a service worker client, we go to network process.
Previously, we were directly going to the client web process.

We are now checking whether the client is a being created client.
A being created client is a client registered in network process when receiving a navigation or main script load.
The client, when fully created, is updating its data by registering itself through IPC.
At this point, the client is no longer a being created client.
A client is a being created client if SWServer:: m_clientPendingMessagesById has an entry with the client ID.

As long as the client is a being created client, we store the message until the client is created in the web process.
When the client is registering itself, we remove the entry in SWServer:: m_clientPendingMessagesById and send all pending messages.

This allows direct support of workers since we reuse the normal post message process.
This also gets us closer to spec, in  particular:
- We ensure message ordering is always preserved.
- We ensure we do not overbuffer messages.

Covered by rebased newly added WPT tests.

* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/controlled-dedicatedworker-postMessage.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/controlled-iframe-postMessage.https-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::m_isNonRenderedPlaceholder):
* Source/WebCore/workers/service/SWClientConnection.cpp:
(WebCore::postMessageToContainer):
(WebCore::SWClientConnection::postMessageToServiceWorkerClient):
* Source/WebCore/workers/service/SWClientConnection.h:
* Source/WebCore/workers/service/ServiceWorkerContainer.cpp:
(WebCore::ServiceWorkerContainer::ServiceWorkerContainer):
(WebCore::ServiceWorkerContainer::startMessages):
(WebCore::ServiceWorkerContainer::postMessage):
(WebCore::ServiceWorkerContainer::addEventListener):
* Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
(WebCore::WorkerSWClientConnection::getServiceWorkerClientPendingMessages): Deleted.
* Source/WebCore/workers/service/WorkerSWClientConnection.h:
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::registerServiceWorkerClient):
(WebCore::SWServer::postMessageToServiceWorkerClient):
(WebCore::SWServer::addServiceWorkerClientPendingMessage): Deleted.
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
(WebKit::WebSWServerConnection::controlClient):
(WebKit::WebSWServerConnection::postMessageToServiceWorkerClient):
(WebKit::WebSWServerConnection::registerServiceWorkerClient):
(WebKit::WebSWServerConnection::registerServiceWorkerClientInternal):
(WebKit::WebSWServerConnection::getServiceWorkerClientPendingMessages): Deleted.
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.messages.in:
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp:
(WebKit::WebSWClientConnection::getServiceWorkerClientPendingMessages): Deleted.
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.h:
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.messages.in:

Canonical link: https://commits.webkit.org/262818@main
